File: modules/inventory/grid.lua
local Items = require 'modules.items.server'

local GRID_COLS = 7
local HOTBAR_SIZE = 7

local TETRIS_TYPES = {
    player = true,
    drop = true,
    newdrop = true,
    container = true,
    stash = true,
    trunk = true,
    glovebox = true,
}

local function isExcludedStash(invId)
    return type(invId) == 'string' and invId:find('^workbench:') ~= nil
end

local Grid = {}

---@param inventory table 
---@return boolean
function Grid.isTetrisType(inventory)
    return TETRIS_TYPES[inventory.type] == true and not isExcludedStash(inventory.id)
end

---@param itemName string
---@return integer width, integer height
function Grid.getItemSize(itemName)
    local item = Items(itemName)
    local grid = item and item.grid
    if not grid then return 1, 1 end
    return grid.width or 1, grid.height or 1
end

---@param inventory table 
---@param slot number
---@return boolean
function Grid.isTetrisSlot(inventory, slot)
    if not Grid.isTetrisType(inventory) then return false end
    if inventory.type == 'player' and slot <= HOTBAR_SIZE then return false end
    return true
end

---@param slot number
---@param w integer
---@param h integer
---@return number[]?
function Grid.getOccupiedCells(slot, w, h)
    if w <= 1 and h <= 1 then return { slot } end

    local col = (slot - 1) % GRID_COLS
    local row = (slot - 1 - col) / GRID_COLS

    if col + w > GRID_COLS then return nil end

    local cells = {}
    local n = 0

    for dy = 0, h - 1 do
        for dx = 0, w - 1 do
            n = n + 1
            cells[n] = (row + dy) * GRID_COLS + (col + dx) + 1
        end
    end

    return cells
end


---@param inventory table server-side Inventory object
---@param anchorSlot number
---@param w integer
---@param h integer
---@param ignoreSlot number
---@return boolean fits
function Grid.canPlace(inventory, anchorSlot, w, h, ignoreSlot)
    if not Grid.isTetrisSlot(inventory, anchorSlot) then

        local occupant = inventory.items[anchorSlot]
        return not occupant or anchorSlot == ignoreSlot
    end

    local cells = Grid.getOccupiedCells(anchorSlot, w, h)
    if not cells then return false end

    for i = 1, #cells do
        if cells[i] > inventory.slots then return false end
    end

    for cellSlot, occupiedItem in pairs(inventory.items) do
        if cellSlot ~= ignoreSlot and Grid.isTetrisSlot(inventory, cellSlot) then
            local ow, oh = Grid.getItemSize(occupiedItem.name)
            local otherCells = Grid.getOccupiedCells(cellSlot, ow, oh)

            if otherCells then
                for i = 1, #otherCells do
                    for j = 1, #cells do
                        if otherCells[i] == cells[j] then
                            return false
                        end
                    end
                end
            end
        end
    end

    return true
end

return Grid

File: web/src/helpers/grid.ts
import { Items } from "../store/items";
import { Slot } from "../typings";

/*
  Tetris-style grid placement — mirrors modules/inventory/grid.lua exactly.
  Used to render items spanning multiple cells and to preview whether a
  drag would fit *before* dropping — this file is UX only. The server
  (modules/inventory/grid.lua) is what actually decides whether a move is
  allowed; nothing here should be treated as authoritative.

  IMPORTANT: GRID_COLS and HOTBAR_SIZE must stay in sync with both
  InventoryGrid.tsx and modules/inventory/grid.lua.
*/

export const GRID_COLS = 7;
export const HOTBAR_SIZE = 7;

const TETRIS_TYPES = new Set([
  "player",
  "drop",
  "newdrop",
  "container",
  "stash",
  "trunk",
  "glovebox",
]);

// The 3x3 crafting workbench is, under the hood, a personal 'stash' named
// 'workbench:<identifier>' — it must stay a plain 1-slot-per-cell grid
// regardless of that, so it's excluded here by id prefix rather than type.
const isExcludedStash = (invId: string) => invId.startsWith("workbench:");

export const isTetrisType = (invType: string, invId: string) =>
  TETRIS_TYPES.has(invType) && !isExcludedStash(invId);

export const isTetrisSlot = (invType: string, invId: string, slot: number) => {
  if (!isTetrisType(invType, invId)) return false;
  if (invType === "player" && slot <= HOTBAR_SIZE) return false;
  return true;
};

export const getItemSize = (
  itemName: string | undefined,
): [width: number, height: number] => {
  const grid = itemName ? Items[itemName]?.grid : undefined;
  return [grid?.width || 1, grid?.height || 1];
};

// Cells (slot numbers) an item of size w x h anchored at `slot` would
// occupy. Returns null if that footprint would run off the right edge of
// the row it starts on.
export const getOccupiedCells = (
  slot: number,
  w: number,
  h: number,
): number[] | null => {
  if (w <= 1 && h <= 1) return [slot];

  const col = (slot - 1) % GRID_COLS;
  const row = (slot - 1 - col) / GRID_COLS;

  if (col + w > GRID_COLS) return null;

  const cells: number[] = [];

  for (let dy = 0; dy < h; dy++) {
    for (let dx = 0; dx < w; dx++) {
      cells.push((row + dy) * GRID_COLS + (col + dx) + 1);
    }
  }

  return cells;
};

// Every cell currently claimed by any item's footprint in `items`, keyed
// by slot number -> the anchor slot of the item covering it. Build this
// once per render/hover check rather than recomputing per-cell.
export const buildOccupancyMap = (
  items: Slot[],
  invType: string,
  invId: string,
  ignoreAnchor?: number,
): Map<number, number> => {
  const occupancy = new Map<number, number>();

  for (const item of items) {
    if (!item.name || item.slot === ignoreAnchor) continue;
    if (!isTetrisSlot(invType, invId, item.slot)) continue;

    const [w, h] = getItemSize(item.name);
    const cells = getOccupiedCells(item.slot, w, h);
    if (!cells) continue;

    for (const cell of cells) occupancy.set(cell, item.slot);
  }

  return occupancy;
};

// Whether an item of size w x h can be placed at `anchorSlot`, given a
// precomputed occupancy map (see buildOccupancyMap) and the inventory's
// total slot count.
export const canPlace = (
  anchorSlot: number,
  w: number,
  h: number,
  occupancy: Map<number, number>,
  totalSlots: number,
): boolean => {
  const cells = getOccupiedCells(anchorSlot, w, h);
  if (!cells) return false;

  for (const cell of cells) {
    if (cell > totalSlots) return false;
    if (occupancy.has(cell)) return false;
  }

  return true;
};

// First anchor slot where an item of size w x h fits — used for
// "auto-place" flows (split-stack modal, double-click, etc) instead of the
// plain "first empty numbered slot" logic that only makes sense pre-tetris.
export const findFirstFit = (
  w: number,
  h: number,
  occupancy: Map<number, number>,
  totalSlots: number,
  startSlot: number,
): number | null => {
  for (let slot = startSlot; slot <= totalSlots; slot++) {
    if (canPlace(slot, w, h, occupancy, totalSlots)) return slot;
  }
  return null;
};
